### ISSUE TITLE
Consider replacing remote execution service with MicroPython-in-WASM

### ISSUE THREAD

**Issue by SnoopJ** [sopel-irc/sopel-py#1]
## tl;dr

what if we replaced the web API for evaluation of small Python programs with MicroPython running on WebAssembly?

---

In July, I wrote a [port of MicroPython to WASI](https://github.com/SnoopJ/micropython/tree/feature/wasi-build/ports/wasi) based on a [WASM + Emscripten port](https://github.com/micropython/micropython/tree/ff70bd1197e581cc43d5a1880323e71676a4b30d/ports/webassembly) that was written as a technical preview by the PyScript folks, and I have been running this port in a custom plugin to my own Sopel instance (:wave: `terribot`) since then. This was just something I did as a challenge to myself, and I have not submitted it upstream so it's _entirely unofficial_, but it has been working very well for a few months now.

To briefly state the value proposition: running a local interpreter gives us more control over what kind of programs can be run and how data is exchanged back to Sopel, and results can be delivered marginally faster without waiting for the network and remote server. Also, Python 3.

I'm filing this issue after a semi-request on IRC by @dgw who has seen the feature running in `terribot`

> 16:01 <+dgw> !py from random import choice; choice([1,2,3])
> 16:01 <+terribot> 3
> ...
> 16:03 <+dgw> SnoopJ, patch for sopel-py when :D
> 16:03 <+dgw> or is it sufficiently tricky to get the MP runtime going that that's impractical?
> 16:03 <+SnoopJ> dgw, maybe sometime. I could packagize it for Linux relatively easily
> 16:04 <+SnoopJ> (it's "just" a ~1 MB .so)

I don't yet feel ready to prepare a PR for `sopel-py`, but it was something I had in mind when I first started dreaming about what I might use a tiny embedded Python for if not for microcontrollers, so this issue is the holding place for intent to work towards a PR.

## Remaining Concerns

Thinking aloud, here are things on my mind between where things stand today and some hypothetical future where this plugin uses the MicroPython port I authored. Feedback/additional items are very welcome in this section, please speak freely.

### Security 

The biggest concern for such a plugin is, of course, the security of the host system and Sopel instance. We want to prevent the execution of arbitrary code in the interpreter where Sopel is running, while also allowing the plugin user to run arbitrary(ish) computations, as well as anything that would be disruptive to the rest of the Sopel instance's business (i.e. obnoxious users should not be able to DoS the bot by evaluating Ackermann function or [insert expensive function reference here]). Users of the plugin **should not be able to:**

* Execute arbitrary Python code in the Python interpreter running Sopel (or otherwise arbitrary code)
* Inject bad data/code into the running Sopel instance or IRC protocol stream
* Meaningfully delay the operation of the running Sopel instance
* Access a file system¹
* Access the network
* Probably other stuff I'm not really listing here

I've had some relatively tech-savvy users hammer on the implementation, and the only thing that was found was a benign bug in the `subprocess` functionality that starts up the guest MicroPython interpreter. The majority of the security of this approach comes from running the guest interpreter inside of WebAssembly using the [WebAssembly System Interface (WASI)](https://github.com/WebAssembly/WASI/) with only the desired system functionalities exposed to the ad-hoc WASM runtime, as well as some feature flags in MicroPython that disable some unnecessary functionality.

All that said, I would probably feel better if we slapped a suitable warning sticker on the plugin if the guest interpreter will be running on user systems, but I guess the EFLv2 covers that for this project anyway.

¹I have Ideas™ about a VFS hosted entirely inside of an ephemeral WASM container, but they aren't appropriate for a first pass at this functionality

### Packaging

Maybe it makes sense to make a dedicated package for the MicroPython-y bits, and update this plugin to use that package, I haven't really decided how I feel about that yet.

Right now this process is still a toy that gets built from my fork of the MicroPython source by a relatively ephemeral build environment, but it shouldn't be too hard to whack up some tooling that will consume that source code and build a `.whl` suitable for Linux systems (using the `manylinux` standard). I am not particularly interested in targeting Windows at the moment but my port could perhaps be tweaked to use the appropriate system functionality on that platform where necessary.

Either way, everything this plugin needs from the C side can be provided in the form of a single shared library (`.so`)

### Standard Library

[MicroPython's standard set of libraries](https://docs.micropython.org/en/latest/library/index.html#python-standard-libraries-and-micro-libraries) is pretty idiosyncratic and surprising to users who are most familiar with the CPython standard library. This is mostly because MicroPython was designed (and is maintained) with microcontrollers in mind, and the libraries are definitely sized down to accomodate for this.

Users will want an experience that is a little more familiar to them from CPython or PyPy, but luckily MicroPython is quite a good implementation, and should probably handle running with a fork of the full upstream standard library just fine (as long as the space allowed for the MicroPython heap is set accordingly), or at least parts of it.

**Comment by dgw** [comment]
Don't have much meaningful to add other than what has already been summarized above, yet. Except that this is an alternative to the (still private and WIP) [cloud function implementation](https://github.com/sopel-irc/pylambda) that I started working on. It too uses Python 3 instead of 2.7, but like the current `oblique` endpoint it maintains separation from the running Sopel instance by virtue of existing as a cloud service.

WASM isn't at all within my field of experience yet 😅 but I do somewhat prefer the elegance of just loading and calling a shared library, vs. something heavier; prior to @SnoopJ's experiments, the best idea I could think of for a local `.py` executor was a Docker/podman/etc. container (still running a pylambda-like implementation for isolating requests from each other, but on localhost instead of a remote provider like AWS/Google Cloud/Cloudflare Workers).
